### DISCUSSION TITLE
How do I override fold queries

### DISCUSSION THREAD

**Discussion by Hubro** [nvim-treesitter/nvim-treesitter#1513] [Q&A] [answered]
I generally think the default folding queries in nvim-treesitter are too aggressive. Everything is folded, down to the smallest `if` block.

I want to override the folding queries to *only* fold classes and functions, but so far the only way I've found to accomplish this is by modifying the original query in place. I don't like this approach since my changes will be overwritten the next time I update the plugin.

I have tried copying the query file to my runtime directory and modifying it, but it doesn't seem to matter where I place it. The query that comes bundled with nvim-treesitter always takes precedence.

I have tried:

- `$HOME/.vim/queries/bash/folds.scm`
- `$HOME/.vim/after/queries/bash/folds.scm`
- `$HOME/.config/nvim/after/queries/bash/folds.scm`
- `$HOME/.config/nvim/after/queries/bash/folds.scm`

In all cases, the original query takes precedence over my custom query.

Original, bundled query file:

```scala
[
  (function_definition)
  (if_statement)
  (case_statement)
  (for_statement)
  (while_statement)
] @fold
```

My override:

```scala
[
  (function_definition)
] @fold
```

What am I doing wrong?

---

(I very recently switched to nvim so I'm using [the config layout suggested here](https://neovim.io/doc/user/nvim.html#nvim-from-vim).)

**Comment by theHamsta** [comment, marked as answer]
You cann't overwrite with runtime files at the moment since the current behavior is to just glue all found query files together. I'm working on a upstream PR to improve the situation (https://github.com/neovim/neovim/pull/14932)

In Lua you can do (to overwrite `fold` queries for Python)
```lua
      if require("nvim-treesitter.parsers").has_parser "python" then
        local folds_query = [[
  [
    (function_definition)
    (class_definition)

    (while_statement)
    (for_statement)
    (if_statement)
    (with_statement)
    (try_statement)

    (import_from_statement)
    (parameters)
    (argument_list)

    (parenthesized_expression)
    (generator_expression)
    (list_comprehension)
    (set_comprehension)
    (dictionary_comprehension)

    (tuple)
    (list)
    (set)
    (dictionary)

    (string)
  ] @fold
  ]]
        require("vim.treesitter.query").set_query("python", "folds", folds_query)
      end
```
You would also be able to read your file from disk and pass the string to the function.

**Reply by andykais** [reply]
Thank you for the answer this is so awesome! I just set one up for typescript to fold on interfaces,classes and class methods too!
```lua
treesitter_parsers = require('nvim-treesitter.parsers')

if treesitter_parsers.has_parser "typescript" then
  local folds_query = [[
    [
      (function_declaration)
      (class_declaration)
      (interface_definition)
      (method_definition)
    ] @fold
  ]]

  require("vim.treesitter.query").set("typescript", "folds", folds_query)
end
```

If I had any suggestion here, it would be to make this way more visible in the docs. I had to hunt down this discussion when I was writing up my own similar question.

**Reply by wookayin** [reply]
Has the behavior changed (in order to invalidate and override the existing query files from nvim-treesitter with user-custom ones) -- since 0.8? Are runtime queries file still *concatenated* instead of overriding? It looks like runtime queries are no longer concatenated but working in an 'overriding' way.

**Reply by wookayin** [reply]
Yes, `vim.treesitter.query.get_files` has changed since 0.8.0 (Sep 2022): https://github.com/neovim/neovim/pull/20117

So it no longer "concatenates" all the query files found in the runtime. Only the first one will be used (unless `; extend` is used.)

See https://github.com/nvim-treesitter/nvim-treesitter/pull/5596 for updating the docs.

**Comment by stsewd** [comment]
You can also set the `foldnestmax` option to a lower value (20 by default), see `:h 'foldnestmax'`
